code/modules/holodeck/basketball.dm

```diff
@@ -0,0 +1,107 @@
+*			unsuccessful: 5 exp; 50 stamina damage; ball goes in a random direction one step away
+*			sucessful: 20 exp; 40-10 stamina damage (based on ball_skill); ball goes in net
+*/
+/obj/structure/hoop/attackby(obj/item/dunk_object, mob/user, params)
+	if(get_dist(src,user)>=2) //can only dunk when close
+		return
+
+	var/ball_skill = user.mind?.get_skill_modifier(/datum/skill/balling, SKILL_PROBS_MODIFIER)
+	var/ball_exp = 0
+
+	if(prob(60 - ball_skill)) //60% base chance to fail a dunk
+		if(isliving(user))
+			var/mob/living/living_baller = user
+			living_baller.apply_damage(50, STAMINA) //failing a dunk is universally taxing
+		ball_exp = 5
+		user?.mind.adjust_experience(/datum/skill/balling, ball_exp)
+		playsound(src, 'sound/weapons/gun/general/grenade_launch.ogg', 50, TRUE)
+		visible_message(span_warning("[user] fumbles the dunk!"))
+		user.dropItemToGround(dunk_object)
+		step(dunk_object, pick(NORTH,SOUTH,EAST,WEST))
+		return
+
+	if(user.transferItemToLoc(dunk_object, drop_location()))
+		var/stam_cost = 40 - ball_skill / 2
+		if(isliving(user))
+			var/mob/living/living_baller = user
+			living_baller.apply_damage(stam_cost, STAMINA)
+		ball_exp = 20
+		user?.mind.adjust_experience(/datum/skill/balling, ball_exp)
+		playsound(src, 'sound/weapons/thudswoosh.ogg', 130, TRUE)
+		visible_message(span_boldnicegreen("[user] dunks [dunk_object] into \the [src]!"))
+
+/obj/structure/hoop/attack_hand(mob/living/user, list/modifiers)
+	. = ..()
+	if(.)
+		return
+	if(user.pulling && isliving(user.pulling))
+		var/mob/living/victim = user.pulling
+		if(user.grab_state < GRAB_AGGRESSIVE)
+			to_chat(user, span_warning("You need a better grip to do that!"))
+			return
+		victim.forceMove(loc)
+		victim.Paralyze(100)
+		visible_message(span_danger("[user] dunks [victim] into \the [src]!"))
+		user.stop_pulling()
+
+/**
+* this proc is for when a hoop is hit by a thrown item, and calculates whether it goes in
+*
+* baller is the person who threw the ball, ball_skill is balling skill level modifier,
+* shot_difficulty is a measure of how difficult a shot is. based on distance to net - (ball_skill / 10). capped between 0 and 9
+* stam_cost is how taxing a shot is. success is not a factor and scales with distance. capped between 5 and 15 stam damage
+* ball_exp is how much exp you get for a shot. success is a factor and scales with distance. capped between 10 and 50
+* odds of *not* making a shot are calculated by taking the prob of shot_difficulty*10. so higher difficulty = harder shot.
+*			unsuccessful: 5 exp; 5-15 stam_cost; ball bounces off
+*			successful: 10-50 exp; 5-15 stam_cost; ball goes in net
+*/
+/obj/structure/hoop/hitby(atom/movable/thrownitem, skipcatch, hitpush, blocked, datum/thrownthing/throwingdatum)
+	if(!isitem(thrownitem) || istype(thrownitem,/obj/projectile))
+		return ..()
+
+	var/mob/baller = throwingdatum?.thrower
+	var/ball_skill = baller.mind?.get_skill_modifier(/datum/skill/balling, SKILL_PROBS_MODIFIER)
```

Comment:
same here, it might be worthwhile to nullcheck these two on the offchance they're null